src/properties/forest.c

```diff
@@ -0,0 +1,216 @@
+                }
+                /*to check for loop in undirected graph*/
+                else if(v==u){
+                    *res=0;
+                    break;
+                }
+            }
+            else{
+                IGRAPH_CHECK(igraph_stack_int_push(&stack, v));
+            }
+        }
+    }
+    igraph_stack_int_destroy(&stack);
+    IGRAPH_FINALLY_CLEAN(1);
+
+    return IGRAPH_SUCCESS;
+}
+
+/**
+ * \ingroup structural
+ * \function igraph_is_forest
+ * \brief Decides whether the graph is a forest.
+ *
+ * An undirected graph is a forest if it has no cycles.
+ * </para><para>
+ *
+ * In the directed case, a possible additional requirement is that a vertex 
+ * should not have more than one parent. Also, edges in each
+ * tree are oriented away from the root (out-tree or arborescence) or all edges
+ * are oriented towards the root (in-tree or anti-arborescence).
+ * This test can be controlled using the \p mode parameter.
+ * </para><para>
+ *
+ * By convention, the null graph (i.e. the graph with no vertices) is considered to be a forest.
+ *
+ * \param graph The graph object to analyze.
+ * \param res Pointer to a logical variable, the result will be stored
+ *        here.
+ * \param roots A vector in which the root nodes will be stored. When \p mode
+ *        is \c IGRAPH_ALL or the graph is undirected, any 1 vertex from each
+ *        component can be the root. When \p mode is \c IGRAPH_OUT
+ *        or \c IGRAPH_IN, all the vertices with zero in- or out-degree,
+ *        respectively are considered as root nodes.
+ * \param mode For a directed graph this specifies whether to test for an
+ *        out-tree, an in-tree or ignore edge directions. The respective
+ *        possible values are:
+ *        \c IGRAPH_OUT, \c IGRAPH_IN, \c IGRAPH_ALL. This argument is
+ *        ignored for undirected graphs.
+ * \return Error code:
+ *        \c IGRAPH_EINVMODE: invalid mode argument.
+ *
+ * Time complexity: At most O(|V|+|E|), the
+ * number of vertices plus the number of edges in the graph.
+ *
+ */
+igraph_error_t igraph_is_forest(const igraph_t *graph,igraph_bool_t *res, igraph_vector_t *roots , igraph_neimode_t mode) {
+    igraph_adjlist_t al;
+    igraph_vector_t visited;
+    igraph_integer_t visited_count=0;
+    igraph_integer_t vcount, ecount;
+
+    vcount = igraph_vcount(graph);
+    ecount = igraph_ecount(graph);
+    IGRAPH_CHECK(igraph_vector_resize(roots,0));
+
+    /*By convention, a zero-vertex graph will be considered a forest.*/
+    if (vcount == 0) {
+        *res = 1;
+        return IGRAPH_SUCCESS;
+    }
+    // A forest can have maximum vcount-1 edges.
+    if (ecount > vcount - 1) {
+        *res = 0;
+        return IGRAPH_SUCCESS;
+    }
+    /*A single-vertex graph is a forest, provided it has no edges (checked in the previous if (..)) */
+    if (vcount == 1) {
+        *res = 1;
+        igraph_vector_push_back(roots,0);
```

Comment:
Please add an `IGRAPH_CHECK()` here. It is very unlikely that the user won't have free memory for pushing one more item to the roots vector, but it may still happen and it's best if you get into the habit of wrapping all igraph function calls in `IGRAPH_CHECK()`.